ir/src/boundary_stmts.rs

```diff
@@ -0,0 +1,201 @@
+    }
+
+    fn last_mut(&mut self) -> &mut BoundaryConstraintsMap {
+        &mut self.last
+    }
+}
+
+impl BoundaryStmts {
+    /// Creates a new [BoundaryStmts] instance with the specified number of trace segments.
+    pub fn new(num_trace_segments: usize) -> Self {
+        Self {
+            boundary_constraints: vec![SegmentBoundaryConstraints::new(); num_trace_segments],
+        }
+    }
+
+    // --- ACCESSORS ------------------------------------------------------------------------------
+
+    /// Returns the number of boundary constraints for the specified trace segment.
+    pub fn num_boundary_constraints(&self, trace_segment: TraceSegment) -> usize {
+        if self.boundary_constraints.len() <= trace_segment.into() {
+            0
+        } else {
+            self.boundary_constraints[trace_segment as usize]
+                .first
+                .len()
+                + self.boundary_constraints[trace_segment as usize].last.len()
+        }
+    }
+
+    /// Returns a vector of the boundary constraints for the specified trace segment at the
+    /// first row.
+    pub fn first_boundary_constraints(
+        &self,
+        trace_segment: TraceSegment,
+    ) -> Vec<(usize, &BoundaryExpr)> {
+        if self.boundary_constraints.len() <= trace_segment.into() {
+            Vec::new()
+        } else {
+            self.boundary_constraints[trace_segment as usize]
+                .first()
+                .iter()
+                .map(|(k, v)| (*k, v))
+                .collect()
+        }
+    }
+
+    /// Returns a vector of the boundary constraints for the specified trace segment at the
+    /// last row.
+    pub fn last_boundary_constraints(
+        &self,
+        trace_segment: TraceSegment,
+    ) -> Vec<(usize, &BoundaryExpr)> {
+        if self.boundary_constraints.len() <= trace_segment.into() {
+            Vec::new()
+        } else {
+            self.boundary_constraints[trace_segment as usize]
+                .last()
+                .iter()
+                .map(|(k, v)| (*k, v))
+                .collect()
+        }
+    }
+
+    // --- MUTATORS -------------------------------------------------------------------------------
+
+    /// Adds a boundary statement from the AST. The statement can either be a variable or a
+    /// constraint. In case it is a constraint, it is added to the boundary_constraints vector
+    /// in the relevant trace segment. Variables are currently not supported for boundary
+    /// constraints.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The identifier specified for the boundary constraint column has not been declared or has
+    ///   been declared with the wrong type.
+    /// - The constraint expression contains invalid public input references.
+    /// - A boundary constraint has already been declared for the specified column and boundary.
+    pub(super) fn insert(
+        &mut self,
+        symbol_table: &mut SymbolTable,
```

Comment:
nit: this shouldn't need to be mutable, since we're not supporting variables in boundary statements yet